drivers/soft_spi/soft_spi.c

```diff
@@ -0,0 +1,207 @@
+{
+    unsigned int spi_num = (unsigned int) bus;
+
+    if (sizeof(spi_config) / sizeof(spi_config[0]) < spi_num || spi_num < 0) {
+        return false;
+    }
+    return true;
+}
+
+void spi_init(spi_t bus)
+{
+    DEBUG("Soft SPI init\n");
+    /* This is left empty because the periph_init function will call this for
+       number of hardware spi ports */
+    assert(spi_bus_is_valid(bus));
+
+    /* initialize device lock */
+    mutex_init(&locks[bus]);
+    spi_init_pins(bus);
+}
+
+void spi_init_pins(spi_t bus)
+{
+    DEBUG("Soft SPI spi_init_pins\n");
+
+    assert(spi_bus_is_valid(bus));
+
+    /* check that miso is not mosi is not clk*/
+    assert(soft_spi_config[bus].mosi_pin != soft_spi_config[bus].miso_pin);
+    assert(soft_spi_config[bus].mosi_pin != soft_spi_config[bus].clk_pin);
+    assert(soft_spi_config[bus].miso_pin != soft_spi_config[bus].clk_pin);
+    /* mandatory pins */
+    assert((GPIO_UNDEF != soft_spi_config[bus].mosi_pin) || (GPIO_UNDEF != soft_spi_config[bus].miso_pin));
+    assert(GPIO_UNDEF != soft_spi_config[bus].clk_pin);
+
+    /* initialize clock pin */
+    gpio_init(soft_spi_config[bus].clk_pin, GPIO_OUT);
+    /* initialize optional pins */
+    if (GPIO_UNDEF != soft_spi_config[bus].mosi_pin) {
+        gpio_init(soft_spi_config[bus].mosi_pin, GPIO_OUT);
+        gpio_clear(soft_spi_config[bus].mosi_pin);
+    }
+}
+
+int spi_init_cs(spi_t bus, spi_cs_t cs)
+{
+    DEBUG("Soft SPI init CS\n");
+    if (!spi_bus_is_valid(bus)) {
+        DEBUG("Soft SPI bus not valid\n");
+        return SPI_NODEV;
+    }
+
+    if (cs == GPIO_UNDEF) {
+        if (soft_spi_config[bus].cs_pin == GPIO_UNDEF) {
+            DEBUG("Soft SPI no valid CS pin\n");
+            return SPI_NOCS;
+        }
+        DEBUG("Soft SPI set default CS line\n");
+        gpio_init(soft_spi_config[bus].cs_pin, GPIO_OUT);
+        gpio_set(soft_spi_config[bus].cs_pin);
+    }
+    else {
+        DEBUG("Soft SPI set user CS line\n");
+        gpio_init(cs, GPIO_OUT);
+        gpio_set(cs);
+    }
+
+    return SPI_OK;
+}
+
+int spi_acquire(spi_t bus, spi_cs_t cs, spi_mode_t mode, spi_clk_t clk)
+{
+    /* Right now we don't care about speed */
+    (void)clk;
+
+    assert(spi_bus_is_valid(bus));
+
+    /* lock bus */
+    mutex_lock(&locks[bus]);
+
+    if (mode < SPI_MODE_0 || SPI_MODE_3 < mode) {
+        return SPI_NOMODE;
+    }
+    soft_spi_config[bus].spi_mode = mode;
+    switch (mode) {
+        case SPI_MODE_0:
+        case SPI_MODE_1:
+            /* CPOL=0 */
+            gpio_clear(soft_spi_config[bus].clk_pin);
+            break;
+        case SPI_MODE_2:
+        case SPI_MODE_3:
+            /* CPOL=1 */
+            gpio_set(soft_spi_config[bus].clk_pin);
+            break;
+    }
+    return SPI_OK;
+}
+
+void spi_release(spi_t bus)
+{
+    assert(spi_bus_is_valid(bus));
+    mutex_unlock(&locks[bus]);
+}
+
+uint8_t _transfer_one_byte(spi_t bus, spi_cs_t cs, bool cont, uint8_t out)
```

Comment:
`static` missing. Better: use `static inline` for this function.

Reply:
✓